moveit_core/online_signal_smoothing/test/test_acceleration_filter.cpp

```diff
@@ -0,0 +1,196 @@
+  EXPECT_FALSE(filter.initialize(node, robot_model_, 3));
+
+  // fail because the update_rate parameter is not set
+  EXPECT_FALSE(filter.initialize(node, robot_model_, 7));
+  node->declare_parameter<double>("update_rate", 0.01);
+
+  // fail because the robot does not have acceleration limits
+  set_limit({});
+  EXPECT_FALSE(filter.initialize(node, robot_model_, 7));
+
+  // succeed with acceleration limits
+  Eigen::VectorXd acceleration_limit = 1.2 * Eigen::VectorXd::Ones(7);
+  set_limit(acceleration_limit);
+  EXPECT_TRUE(filter.initialize(node, robot_model_, 7));
+}
+
+TEST_F(AccelerationFilterTest, FilterDoSmooth)
+{
+  online_signal_smoothing::AccelerationLimitedPlugin filter;
+
+  rclcpp::Node::SharedPtr node = std::make_shared<rclcpp::Node>("AccelerationFilterTest");
+  node->declare_parameter<std::string>("move_group_name", move_group_name_);
+  double update_rate = 1.0;
+  node->declare_parameter<double>("update_rate", update_rate);
+  Eigen::VectorXd acceleration_limit = 1.2 * Eigen::VectorXd::Ones(7);
+  set_limit(acceleration_limit);
+
+  EXPECT_TRUE(filter.initialize(node, robot_model_, 7));
+
+  // fail when called with the wrong number of joints
+  Eigen::VectorXd position = Eigen::VectorXd::Zero(5);
+  Eigen::VectorXd velocity = Eigen::VectorXd::Zero(5);
+  Eigen::VectorXd acceleration = Eigen::VectorXd::Zero(5);
+  EXPECT_FALSE(filter.doSmoothing(position, velocity, acceleration));
+
+  // fail because reset was not called
+  position = Eigen::VectorXd::Zero(7);
+  velocity = Eigen::VectorXd::Zero(7);
+  acceleration = Eigen::VectorXd::Zero(7);
+  EXPECT_FALSE(filter.doSmoothing(position, velocity, acceleration));
+
+  // succeed
+  filter.reset(Eigen::VectorXd::Zero(7), Eigen::VectorXd::Zero(7), Eigen::VectorXd::Zero(7));
+  EXPECT_TRUE(filter.doSmoothing(position, velocity, acceleration));
+
+  // succeed: apply acceleration limits
+  filter.reset(Eigen::VectorXd::Zero(7), Eigen::VectorXd::Zero(7), Eigen::VectorXd::Zero(7));
+  position.array() = 3.0;
+  EXPECT_TRUE(filter.doSmoothing(position, velocity, acceleration));
+  EXPECT_TRUE((position.array() - update_rate * update_rate * acceleration_limit.array()).matrix().norm() < 1E-3);
+
+  // succeed: apply acceleration limits
+  position.array() = 0.9;
+  filter.reset(position * 0, velocity * 0, acceleration * 0);
+  EXPECT_TRUE(filter.doSmoothing(position, velocity, acceleration));
+  EXPECT_TRUE((position.array() - 0.9).matrix().norm() < 1E-3);
+
+  // succeed: slow down
+  velocity = 10 * Eigen::VectorXd::Ones(7);
+  filter.reset(Eigen::VectorXd::Zero(7), velocity, Eigen::VectorXd::Zero(7));
+  position.array() = 0.01;
+  Eigen::VectorXd expected_offset =
+      velocity.array() * update_rate - update_rate * update_rate * acceleration_limit.array();
+  EXPECT_TRUE(filter.doSmoothing(position, velocity, acceleration));
+  EXPECT_TRUE((velocity * update_rate - expected_offset).norm() < 1E-3);
+}
+
+TEST_F(AccelerationFilterTest, FilterBadAccelerationConfig)
```

Comment:
Is this bad because the accel limits can't be negative?

Reply:
Yeah if the maximum acceleration is negative or the minimum is positive, then there is no solution to the problem. The robot can’t be controlled.